protocol/parser/parser.go

```diff
@@ -163,6 +237,118 @@ func parse(rpcInput RPCInput, blockParser spectypes.BlockParser, dataSource int)
 	return retval, nil
 }
 
+type ParsedInput struct {
+	parsedBlock  int64
+	parsedHashes []string
+}
+
+func NewParsedInput() *ParsedInput {
+	return &ParsedInput{
+		parsedBlock:  spectypes.NOT_APPLICABLE,
+		parsedHashes: make([]string, 0),
+	}
+}
+
+func (p *ParsedInput) SetBlock(block int64) {
+	p.parsedBlock = block
+}
+
+func (p *ParsedInput) GetBlock() int64 {
+	return p.parsedBlock
+}
+
+func (p *ParsedInput) GetBlockHashes() ([]string, error) {
+	if len(p.parsedHashes) == 0 {
+		return nil, fmt.Errorf("no parsed hashes found")
+	}
+	return p.parsedHashes, nil
+}
+
+func ParseWithGenericParsers(rpcInput RPCInput, genericParsers []spectypes.GenericParser) (*ParsedInput, error) {
```

Comment:
we need to tell the parser what we are trying to parse: in this case blocks and hashes from params, so it knows to skip other parsers such as error and result parsers, you can use a map of allowed parsers by the enum as an argument